File: robert-server-database/src/main/java/fr/gouv/stopc/robertserver/database/repository/RegistrationRepository.java
package fr.gouv.stopc.robertserver.database.repository;

import fr.gouv.stopc.robertserver.database.model.Registration;
import org.springframework.data.mongodb.repository.MongoRepository;
import org.springframework.data.mongodb.repository.Query;
import org.springframework.stereotype.Repository;

@Repository
public interface RegistrationRepository extends MongoRepository<Registration, byte[]> {

    /**
     * Retrieve the number of users at risk
     *
     * @return the count result
     */
    @Query(value = "{ atRisk: {$eq: true} }", count = true)
    Long countNbUsersAtRisk();

    /**
     * Retrieve the number of users than epoch exposition with an epochId <= the
     * epochId in parameter.
     *
     * @param minEpochId
     * @return the count result
     */
    @Query(value = "{exposedEpochs:{$elemMatch:{epochId:{$lte: ?0}}}}", count = true)
    Long countNbUsersWithOldEpochExpositions(int minEpochId);
}

File: robert-server-ws-rest/src/main/kotlin/fr/gouv/stopc/robertserver/ws/repository/RegistrationRepository.kt
package fr.gouv.stopc.robertserver.ws.repository

import fr.gouv.stopc.robertserver.ws.repository.model.Registration
import org.springframework.data.mongodb.core.MongoOperations
import org.springframework.data.mongodb.core.query.Query
import org.springframework.data.mongodb.core.query.Update
import org.springframework.data.mongodb.core.query.isEqualTo
import org.springframework.data.mongodb.repository.MongoRepository
import org.springframework.stereotype.Repository
import java.time.Duration

@Repository
interface RegistrationRepository : MongoRepository<Registration, ByteArray>, RegistrationCustomRepository

interface RegistrationCustomRepository {
    fun updateLastTimestampDrift(idA: ByteArray, driftTime: Duration)
}

class RegistrationCustomRepositoryImpl(private val mongoOperations: MongoOperations) :
    RegistrationCustomRepository {

    override fun updateLastTimestampDrift(idA: ByteArray, driftTime: Duration) {
        mongoOperations.update(Registration::class.java)
            .matching(Query.query(Registration::permanentIdentifier isEqualTo idA))
            .apply(Update.update("lastTimestampDrift", driftTime.seconds))
            .first()
    }
}
